### PR TITLE
Content Packs Cloud Storage

### PR DESCRIPTION
**Pull request by IkaDemisto**
## Status
Ready

## Related Issues
related to: https://github.com/demisto/etc/issues/20091

## Description
Implemented logic of uploading packs to cloud storage. 

## Does it break backward compatibility?
   - No

## Must have
- [ ] Tests
- [ ] Documentation (with link to it)
- [x] Code Review

### CODE REVIEW

File: Tests/store_packs.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import json
+import os
+import sys
+import argparse
+
```
Comment (ronykoz) [inline Tests/store_packs.py]: remove

File: Tests/store_packs.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+from os import path
```
Comment (ronykoz) [inline Tests/store_packs.py]: both this and import os?

File: Tests/store_packs.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        print(f"Uploaded {self.pack_name} pack to {pack_full_path} path.")
+
+
+def get_modified_packs(is_circle=False, specific_pack=None):
+    if not is_circle:
+        return [specific_pack]
+
+    cmd = f"git diff --name-only HEAD..HEAD^ | grep 'Packs/'"
+    modified_packs_path = run_command(cmd, use_shell=True).splitlines()
+    modified_packs = [p.split('/')[1] for p in modified_packs_path]
```
Comment (ronykoz) [inline Tests/store_packs.py]: should be set

File: Tests/store_packs.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+def extract_modified_packs(modified_packs, packs_artifacts_path, extract_destination_path):
+    with ZipFile(packs_artifacts_path) as packs_artifacts:
+        for pack in packs_artifacts.namelist():
+            for modified_pack in modified_packs:
+
+                if pack.startswith(f"{modified_pack}/"):
+                    packs_artifacts.extract(pack, extract_destination_path)
+
+    print(f"Extracted {', '.join(modified_packs)} packs to path: {extract_destination_path}")
```
Comment (ronykoz) [inline Tests/store_packs.py]: print the extract in the loop

File: Tests/store_packs.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    parser.add_argument('-e', '--extractPath', help="Full path of folder to extract wanted packs", required=True)
+    parser.add_argument('-c', '--circleCi', help="Whether run script locally or in circleCi", default=False)
+    parser.add_argument('-p', '--packName', help="Use only in local mode, the target pack name to store.",
+                        required=False, default="")
+    parser.add_argument('-b', '--bucketName', help="Storage bucket name", required=True)
+
+    return parser.parse_args()
+
+
+class PackStorage:
```
Comment (ronykoz) [inline Tests/store_packs.py]: maybe just name it as pack?

File: Tests/store_packs.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        with ZipFile(zip_pack_path, 'w', ZIP_DEFLATED) as pack_zip:
+            for root, dirs, files in os.walk(self.pack_path):
+                for f in files:
+                    full_file_path = path.join(root, f)
+                    relative_file_path = path.relpath(full_file_path, self.pack_path)
+                    pack_zip.write(filename=full_file_path, arcname=relative_file_path)
+
+        return zip_pack_path
+
+    def store_pack(self, zip_pack_path, latest_version):
```
Comment (ronykoz) [inline Tests/store_packs.py]:
```suggestion
    def upload_pack_to_s3(self, zip_pack_path, latest_version):
```

File: Tests/store_packs.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            return self.PACK_INITIAL_VERSION
+
+        changelog_path = path.join(self.pack_path, self.CHANGELOG)
+
+        with open(changelog_path, "r") as changelog:
+            changelog_json = json.load(changelog)
+
+            pack_versions = list(changelog_json.keys())
+            pack_versions.sort(key=lambda str_version: [int(v) for v in str_version.split(".")],
+                               reverse=True)
```
Comment (ronykoz) [inline Tests/store_packs.py]: maybe we should import a package for this, to avoid any errors that may occur?

File: Tests/test_utils.py [hunk 8]
```diff
@@ -35,21 +35,25 @@ def print_warning(warning_str):
...
 
     Returns:
         string. The output of the command you are trying to execute.
     """
+    if not use_shell:
+        command = command.split()
+
     if is_silenced:
-        p = Popen(command.split(), stdout=PIPE, stderr=PIPE, universal_newlines=True)
+        p = Popen(command, stdout=PIPE, stderr=PIPE, universal_newlines=True, shell=use_shell)
```
Comment (ronykoz) [inline Tests/test_utils.py]: should be done in the other if i guess, and i think that it does that by default

File: Tests/store_packs.py (hunk 6 above)
Comment (IkaDemisto) [reply Tests/store_packs.py]: `s3` can be confusing, maybe `upload_pack_to_storage` instead?

File: Tests/store_packs.py (hunk 7 above)
Comment (IkaDemisto) [reply Tests/store_packs.py]: We can use this package https://pypi.org/project/semantic-version/. But I think that we should use it in validation (of semantic versioning) of `changelog.json` instead, which will include all packs versions.

File: Tests/test_utils.py (hunk 8 above)
Comment (IkaDemisto) [reply Tests/test_utils.py]: Need it in order to use pipe commands, check https://stackoverflow.com/questions/13332268/how-to-use-subprocess-command-with-pipes
When the repo will contain only packs, we will no longer need the `| grep 'Packs/'` part.

**Comment by IkaDemisto** [comment]
I will add docstring later

File: Tests/test_utils.py (hunk 8 above)
Comment (ronykoz) [inline Tests/test_utils.py]: but the else option covers it

File: Tests/test_utils.py [hunk 9]
```diff
@@ -35,21 +35,25 @@ def print_warning(warning_str):
...
         command (string): The string of the command you want to execute.
         is_silenced (bool): Whether to print command output.
         exit_on_error (bool): Whether to exit on command error.
+        use_shell (bool): Whether use shell to executed command through the shell.
 
     Returns:
         string. The output of the command you are trying to execute.
     """
+    if not use_shell:
+        command = command.split()
```
Comment (ronykoz) [inline Tests/test_utils.py]: why?

File: Tests/store_packs.py (hunk 7 above)
Comment (ronykoz) [reply Tests/store_packs.py]: 👍

File: Tests/store_packs.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        version_pack_path = f"content/packs/{self.pack_name}/{latest_version}"
+        existing_files = [f.name for f in self.storage_bucket.list_blobs(prefix=version_pack_path)]
+
+        if len(existing_files) > 0:
+            print_error(f"The following packs already exist at storage: {', '.join(existing_files)}")
+            print_error(f"Skipping step of uploading {self.pack_name}.zip to storage.")
+            sys.exit(1)
+
+        pack_full_path = f"{version_pack_path}/{self.pack_name}.zip"
+        blob = self.storage_bucket.blob(pack_full_path)
```
Comment (ronykoz) [inline Tests/store_packs.py]: blob?

Comment (IkaDemisto) [reply Tests/store_packs.py]: Official google storage python client library, they call the object stored in the path blobs

File: Tests/test_utils.py (hunk 8 above)
Comment (IkaDemisto) [reply Tests/test_utils.py]: Didn't get you here

**Comment by IkaDemisto** [comment]
@yaakovi @ronykoz please review and give your feedback, so we can add additional step in `config.yml` with this script.

File: Tests/test_utils.py [hunk 11]
```diff
@@ -342,3 +347,85 @@ def get_pack_name(file_path):
...
+#     integration_dir = os.path.join(pack_path, 'Integrations')
+#     for integration_name in os.listdir(integration_dir):
+#         if integration_name.endswith('.yml'):
+#             integration_yml_path = os.path.join(integration_dir, integration_name)
+#             with open(integration_yml_path, 'r') as integration_file:
+#                 yaml_data = yaml.safe_load(integration_file)
+#                 display_name = yaml_data.get('display', '')
+#                 integration_display_names.add(display_name)
+#
+#     return integration_display_names
```
Comment (ronykoz) [inline Tests/test_utils.py]: remove

**Comment by IkaDemisto** [comment]
@guydemi  FYI

File: Tests/store_packs.py [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+    specific_packs = option.pack_names
+    build_number = option.ci_build_number if option.ci_build_number else str(uuid.uuid4())
+
+    storage_client = init_storage_client(service_account)
+    storage_bucket = storage_client.get_bucket(storage_bucket_name)
+
+    modified_packs = get_modified_packs(specific_packs)
+    extract_modified_packs(modified_packs, packs_artifacts_path, extract_destination_path)
+    packs_list = [Pack(pack_name, os.path.join(extract_destination_path, pack_name), storage_bucket)
+                  for pack_name in modified_packs]
```
Comment (yaakovi) [inline Tests/store_packs.py]: switch between those two.
first do the logics then handle the network stuff

File: Tests/store_packs.py [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+
+DIR_NAME_TO_CONTENT_TYPE = {
+    "Scripts": "Automations",
+    "IncidentFields": "Incident Fields",
+    "Playbooks": "Playbooks",
+    "Integrations": "Integrations",
+    "IncidentTypes": "Incident Types",
+    "Layouts": "Incident Layouts",
+    "Reports": "Reports",
+    "Dashboards": "Dashboards"
```
[the remark this answers is not in the document]
Comment (IkaDemisto) [reply Tests/store_packs.py]: Updated:
{
    "Classifiers": "Classifiers",
    "Dashboards": "Dashboards",
    "IncidentFields": "Incident Fields",
    "IncidentTypes": "Incident Types",
    "IndicatorFields": "Indicator Fields",
    "Integrations": "Integrations",
    "Layouts": "Incident Layouts",
    "Playbooks": "Playbooks",
    "Reports": "Reports",
    "Scripts": "Automations",
    "Widgets": "Widgets"
}

File: .circleci/config.yml [hunk 14]
```diff
@@ -441,6 +441,18 @@ jobs:
...
           when: always
+      - run:
+          name: Upload Changed Packs To Storage
+          command: |
+            if [[ $CIRCLE_BRANCH != master ]]; then
+              echo "Skipping packs uploading on non master branch"
+              exit 0
+            fi
+
+            EXTRACT_FOLDER=$(mktemp -d)
```
[the remark this answers is not in the document]
Comment (IkaDemisto) [reply .circleci/config.yml]: very nice

File: Tests/store_packs.py [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+
+def extract_modified_packs(modified_packs, packs_artifacts_path, extract_destination_path):
+    print("Starting extracting modified pack:")
+    with ZipFile(packs_artifacts_path) as packs_artifacts:
+        for pack in packs_artifacts.namelist():
+            for modified_pack in modified_packs:
+
+                if pack.startswith(f"{modified_pack}/"):
+                    packs_artifacts.extract(pack, extract_destination_path)
+                    print(f"Extracted {pack} to path: {extract_destination_path}")
```
[the remark this answers is not in the document]
Comment (IkaDemisto) [reply Tests/store_packs.py]: Can be more than one pack, modified_packs is set with more than one result.

**Comment by IkaDemisto** [comment]
FYI @DeanArbel @amshamah419

**Comment by IkaDemisto** [comment]
@glicht can we have separate bucket for this with service account permissions (same as in `marketplace-dist-dev` bucket)? In order to avoid collision with already uploaded manually packs that server use for testing.

File: Tests/store_packs.py [hunk 16]
```diff
@@ -0,0 +1,512 @@
...
+# global constants
+STORAGE_BASE_PATH = "content/packs"
+IGNORED_PATHS = ['Packs/__init__.py']
+# the format is defined in issue #19786, may change in the future
+DIR_NAME_TO_CONTENT_TYPE = {
+    "Classifiers": "classifier",
+    "Dashboards": "dashboard",
+    "IncidentFields": "incidentfield",
+    "IncidentTypes": "incidenttype",
+    # "IndicatorFields": "indicatorfields",
```
Comment (yuvalbenshalom) [inline Tests/store_packs.py]: @IkaDemisto are we removing this or not?

File: Tests/store_packs.py [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+        METADATA (str): pack's metadata file name, may be changed in the future.
+        INDEX_NAME (str): pack's index name, may be changed in the future.
+
+    """
+    PACK_INITIAL_VERSION = "1.0.0"
+    DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'
+    CHANGELOG_JSON = "changelog.json"
+    CHANGELOG_MD = "changelog.md"
+    README = "README.md"
+    METADATA = "metadata.json"
```
Comment (yuvalbenshalom) [inline Tests/store_packs.py]: is this the me metadata in the content repo or the uploaded one?

File: Tests/store_packs.py [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+        pack_full_path = f"{version_pack_path}/{self._pack_name}.zip"
+        blob = storage_bucket.blob(pack_full_path)
+
+        with open(zip_pack_path, "rb") as pack_zip:
+            blob.upload_from_file(pack_zip)
+        print_color(f"Uploaded {self._pack_name} pack to {pack_full_path} path.", LOG_COLORS.GREEN)
+
+        return True
+
+    def format_metadata(self):
```
Comment (yuvalbenshalom) [inline Tests/store_packs.py]: not sure I understand this func, Need to make sure that our internal metadata name was changed.

File: Tests/store_packs.py (hunk 17 above)
Comment (IkaDemisto) [reply Tests/store_packs.py]: Currently they have same name, we need to change the name of one in repo

File: Tests/store_packs.py (hunk 16 above)
Comment (IkaDemisto) [reply Tests/store_packs.py]: we will include it as `reputation` field.

File: Tests/store_packs.py (hunk 18 above)
Comment (IkaDemisto) [reply Tests/store_packs.py]: Changed internal name

**Comment by IkaDemisto** [comment]
Example of triggered build: https://circleci.com/gh/demisto/content/43199 (task `Upload Packs To Marketplace Storage`)

**Comment by IkaDemisto** [comment]
@yuvalbenshalom test with new demisto-sdk version: https://circleci.com/gh/demisto/content/43402#tests/containers/1
Failed task aren't related to the pr and fail on lint of `CommonServerPython` and `Packs/FeedRecordedFuture/Integrations/FeedRecordedFuture`

**Comment by IkaDemisto** [comment]
with 0.4.2 version: https://circleci.com/gh/demisto/content/43469#tests/containers/1

**Pull request merged**